File: ecmascript/beer-song/beer-song.js
function* range(begin, end) {
    for (let i = begin; i <= end; i++) {
        yield i;
    }
}

export default class BeerSong {
    static verse(number) {
        switch (number) {
            case 0:
                return "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n";
            case 1:
                return "1 bottle of beer on the wall, 1 bottle of beer.\nTake it down and pass it around, no more bottles of beer on the wall.\n";
            case 2:
                return "2 bottles of beer on the wall, 2 bottles of beer.\nTake one down and pass it around, 1 bottle of beer on the wall.\n";
            default:
                return `${number} bottles of beer on the wall, ${number} bottles of beer.\nTake one down and pass it around, ${number - 1} bottles of beer on the wall.\n`;
        }
    }

    static sing(start = 99, stop = 0) {
        const verses = [...range(stop, start)].reverse().map(BeerSong.verse);
        return verses.join("\n");
    }
}

File: v1/kotlin/beer-song/src/main/kotlin/BeerSong.kt
object BeerSong {
    fun verse(number: Int): String {
        if (number < 0 || number > 99)
            throw IllegalArgumentException()

        return when (number) {
            0 -> "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n"
            1 -> "1 bottle of beer on the wall, 1 bottle of beer.\nTake it down and pass it around, no more bottles of beer on the wall.\n"
            2 -> "2 bottles of beer on the wall, 2 bottles of beer.\nTake one down and pass it around, 1 bottle of beer on the wall.\n"
            else -> "$number bottles of beer on the wall, $number bottles of beer.\nTake one down and pass it around, " + (number - 1) + " bottles of beer on the wall.\n"
        }
    }

    fun verses(stop: Int, start: Int): String =
        (stop downTo start)
            .map { verse(it) }
            .joinToString(separator = "\n" )

    val lyrics = verses(99, 0)
}